File: go/test/iframe_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/bluefin-payconex-sdk/go"
	"github.com/voxgig-sdk/bluefin-payconex-sdk/go/core"

	vs "github.com/voxgig-sdk/bluefin-payconex-sdk/go/utility/struct"
)

func TestIframeEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Iframe(nil)
		if ent == nil {
			t.Fatal("expected non-nil IframeEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"iframe": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Iframe(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Iframe(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := iframeBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "iframe." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set BLUEFINPAYCONEX_TEST_IFRAME_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		iframeRef01Ent := client.Iframe(nil)
		iframeRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "iframe"}, setup.data), "iframe_ref01"))
		iframeRef01Data["account_id"] = setup.idmap["account01"]

		iframeRef01DataResult, err := iframeRef01Ent.Create(iframeRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		iframeRef01Data = core.ToMapAny(iframeRef01DataResult)
		if iframeRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if iframeRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		iframeRef01Match := map[string]any{
			"account_id": setup.idmap["account01"],
		}

		iframeRef01ListResult, err := iframeRef01Ent.List(iframeRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		iframeRef01List, iframeRef01ListOk := iframeRef01ListResult.([]any)
		if !iframeRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", iframeRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(iframeRef01List), map[string]any{"id": iframeRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		iframeRef01DataUp0Up := map[string]any{
			"id": iframeRef01Data["id"],
			"account_id": setup.idmap["account_id"],
		}

		iframeRef01MarkdefUp0Name := "currency"
		iframeRef01MarkdefUp0Value := fmt.Sprintf("Mark01-iframe_ref01_%d", setup.now)
		iframeRef01DataUp0Up[iframeRef01MarkdefUp0Name] = iframeRef01MarkdefUp0Value

		iframeRef01ResdataUp0Result, err := iframeRef01Ent.Update(iframeRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		iframeRef01ResdataUp0 := core.ToMapAny(iframeRef01ResdataUp0Result)
		if iframeRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if iframeRef01ResdataUp0["id"] != iframeRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if iframeRef01ResdataUp0[iframeRef01MarkdefUp0Name] != iframeRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", iframeRef01MarkdefUp0Name, iframeRef01ResdataUp0[iframeRef01MarkdefUp0Name])
		}

		// LOAD
		iframeRef01MatchDt0 := map[string]any{
			"id": iframeRef01Data["id"],
		}
		iframeRef01DataDt0Loaded, err := iframeRef01Ent.Load(iframeRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		iframeRef01DataDt0LoadResult := core.ToMapAny(iframeRef01DataDt0Loaded)
		if iframeRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if iframeRef01DataDt0LoadResult["id"] != iframeRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func iframeBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "iframe", "IframeTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read iframe test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse iframe test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"iframe01", "iframe02", "iframe03", "account01", "account02", "account03", "payment_iframe01", "payment_iframe02", "payment_iframe03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("BLUEFINPAYCONEX_TEST_IFRAME_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"BLUEFINPAYCONEX_TEST_IFRAME_ENTID": idmap,
		"BLUEFINPAYCONEX_TEST_LIVE":      "FALSE",
		"BLUEFINPAYCONEX_TEST_EXPLAIN":   "FALSE",
		"BLUEFINPAYCONEX_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["BLUEFINPAYCONEX_TEST_IFRAME_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add account_id alias for update test.
	if idmapResolved["account_id"] == nil {
		idmapResolved["account_id"] = idmapResolved["account01"]
	}

	if env["BLUEFINPAYCONEX_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["BLUEFINPAYCONEX_APIKEY"],
			},
			extra,
		})
		client = sdk.NewBluefinPayconexSDK(core.ToMapAny(mergedOpts))
	}

	live := env["BLUEFINPAYCONEX_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["BLUEFINPAYCONEX_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: cpp/test/iframe_entity_test.cpp
// Generated basic-flow test for the iframe entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct IframeSetup {
  std::shared_ptr<BluefinPayconexSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static IframeSetup iframe_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/iframe/IframeTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinPayconexSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("iframe01"), Value("iframe02"), Value("iframe03"), Value("account01"), Value("account02"), Value("account03"), Value("payment_iframe01"), Value("payment_iframe02"), Value("payment_iframe03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINPAYCONEX_TEST_IFRAME_ENTID", idmap},
    {"BLUEFINPAYCONEX_TEST_LIVE", Value("FALSE")},
    {"BLUEFINPAYCONEX_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINPAYCONEX_TEST_IFRAME_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  if (getp(idmap_resolved, "account_id").is_undef()) {
    setp(idmap_resolved, "account_id", getp(idmap_resolved, "account01"));
  }

  bool live = getp(env, "BLUEFINPAYCONEX_TEST_LIVE") == Value("TRUE");

  IframeSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void iframe_entity_instance() {
  auto testsdk = BluefinPayconexSDK::testSDK();
  auto ent = testsdk->iframe();
  ASSERT_EQ(ent->getName(), std::string("iframe"), "entity name");
}

static void iframe_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"iframe", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinPayconexSDK::testSDK(seed, sdkopts);
  auto se = strsdk->iframe();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinPayconexSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->iframe();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void iframe_entity_basic() {
  auto setup = iframe_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create", "list", "update", "load"}) {
    auto sk = is_control_skipped("entityOp", std::string("iframe.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto iframe_ref01_ent = client->iframe();
  Value iframe_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "iframe"}), "iframe_ref01"));
  if (!iframe_ref01_data.is_map()) iframe_ref01_data = vmap();
  setp(iframe_ref01_data, "account_id", getp(setup.idmap, "account01"));
  {
    Value iframe_ref01_data_result = iframe_ref01_ent->create(Struct::clone(iframe_ref01_data), Value::undef());
    iframe_ref01_data = Helpers::toMapAny(iframe_ref01_data_result);
    if (!iframe_ref01_data.is_map()) iframe_ref01_data = vmap();
    ASSERT_TRUE(iframe_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(iframe_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

  // LIST
  Value iframe_ref01_match = vmap();
  setp(iframe_ref01_match, "account_id", getp(setup.idmap, "account01"));
  Value iframe_ref01_list = iframe_ref01_ent->list(Struct::clone(iframe_ref01_match), Value::undef());
  ASSERT_TRUE(iframe_ref01_list.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(iframe_ref01_list), vmap({{"id", getp(iframe_ref01_data, "id")}}));
    ASSERT_TRUE(!found.empty(), "expected to find created entity in list");
  }

  // UPDATE
  Value iframe_ref01_data_up0_up = vmap();
  setp(iframe_ref01_data_up0_up, "id", getp(iframe_ref01_data, "id"));
  setp(iframe_ref01_data_up0_up, "account_id", getp(setup.idmap, "account_id"));
  std::string iframe_ref01_data_up0_markval = std::string("Mark01-iframe_ref01_") + std::to_string(setup.now);
  setp(iframe_ref01_data_up0_up, "currency", Value(iframe_ref01_data_up0_markval));
  Value iframe_ref01_resdata_up0_result = iframe_ref01_ent->update(Struct::clone(iframe_ref01_data_up0_up), Value::undef());
  Value iframe_ref01_resdata_up0 = Helpers::toMapAny(iframe_ref01_resdata_up0_result);
  if (!iframe_ref01_resdata_up0.is_map()) iframe_ref01_resdata_up0 = vmap();
  ASSERT_TRUE(iframe_ref01_resdata_up0.is_map(), "expected update result to be a map");
  ASSERT_EQ_VAL(getp(iframe_ref01_resdata_up0, "id"), getp(iframe_ref01_data_up0_up, "id"), "expected update result id to match");
  ASSERT_EQ_VAL(getp(iframe_ref01_resdata_up0, "currency"), Value(iframe_ref01_data_up0_markval), "expected currency to be updated");

  // LOAD
  Value iframe_ref01_match_dt0 = vmap({{"id", getp(iframe_ref01_data, "id")}});
  Value iframe_ref01_data_dt0_loaded = iframe_ref01_ent->load(Struct::clone(iframe_ref01_match_dt0), Value::undef());
  Value iframe_ref01_data_dt0_load_result = Helpers::toMapAny(iframe_ref01_data_dt0_loaded);
  ASSERT_TRUE(iframe_ref01_data_dt0_load_result.is_map(), "expected load result to be a map");
  ASSERT_EQ_VAL(getp(iframe_ref01_data_dt0_load_result, "id"), getp(iframe_ref01_data, "id"), "expected load result id to match");

}

int main() {
  T_RUN(iframe_entity_instance);
  T_RUN(iframe_entity_stream);
  T_RUN(iframe_entity_basic);
  return sdktest::summary("iframe_entity_test");
}
